### PR TITLE
Allow extract-dat to extract single subfile to stdout

### LINKED ISSUE #4972: spectre extract-dat - allow extracting to stdout
Allowing extraction of individual dat subfiles to stdout would be useful for working with analysis software that can read from a pipe.

### PR DESCRIPTION
**Pull request by knelli2**
## Proposed changes

Fixes #4972 

Prints something like (for a 1 line dat file)
```
 Time  NumberOfPoints       Volume  L2Norm(GaugeConstraint)  L2Norm(ThreeIndexConstraint)
 0.02        154350.0 4.188761e+06             5.762183e-08                      0.000002
```

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by knelli2** [comment]
@wthrowe Does this seem like it'll do what you want?

File: src/IO/H5/Python/ExtractDatFromH5.py [hunk 1]
```diff
@@ -99,6 +124,11 @@ def extract_dat_files(filename, out_dir, num_cores, list=False, force=False):
               '-l',
               is_flag=True,
               help="List all dat files in the HDF5 file and exit.")
+@click.option('--subfile',
+              '-s',
```
Comment (nilsvu) [inline src/IO/H5/Python/ExtractDatFromH5.py]: `-d` for consistency with other tools

File: src/IO/H5/Python/ExtractDatFromH5.py [hunk 2]
```diff
@@ -25,56 +22,84 @@ def save_dat_data(dat_path, h5_filename, out_dir):
...
-                zip(all_dat_files, [filename] * num_dat_files,
-                    [out_dir] * num_dat_files))
-    else:
-        for dat_filename in all_dat_files:
-            save_dat_data(dat_filename, filename, out_dir)
+    # Ignore multiprocessing if we are only extracting one file to stdout
+    if out_dir:
+        # Only use multiprocessing if we are using more than one core. Otherwise
+        # avoid the overhead
+        if num_cores > 1:
```
Comment (nilsvu) [inline src/IO/H5/Python/ExtractDatFromH5.py]: `out_dir and num_cores > 1`

File: src/IO/H5/Python/ExtractDatFromH5.py [hunk 3]
```diff
@@ -25,56 +22,84 @@ def save_dat_data(dat_path, h5_filename, out_dir):
...
+                    write_dat_data,
+                    zip(all_dat_files, [filename] * num_dat_files,
+                        [out_dir] * num_dat_files))
+        else:
+            for dat_filename in all_dat_files:
+                write_dat_data(dat_filename, filename, out_dir)
 
-    print(f"Successfully extracted all Dat files into '{out_dir}'")
+        print(f"Successfully extracted all Dat files into '{out_dir}'")
+    else:
```
Comment (nilsvu) [inline src/IO/H5/Python/ExtractDatFromH5.py]: Don't need this else clause. The loop above is enough.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: tests/Unit/IO/H5/Python/Test_ExtractDatFromH5.py [hunk 4]
```diff
@@ -93,6 +93,30 @@ def test_extract_dat_files(self):
...
+                          out_dir=None,
+                          num_cores=1,
+                          list=False,
+                          force=False,
+                          subfiles=["TimeSteps2.dat"])
+
+        # We shouldn't have created an out directory
+        self.assertFalse(os.path.exists(self.created_dir))
+
+        self.failUnlessRaises(
```
Comment (nilsvu) [inline tests/Unit/IO/H5/Python/Test_ExtractDatFromH5.py:115]: (optional) This is nicer:

```
with self.assertRaises(ValueError):
    extract...
```

**Review by nilsvu — CHANGES_REQUESTED** [review]

**Review by nilsvu — DISMISSED** [review]

**Comment by wthrowe** [comment]
Is there an option to set the precision?  Your example appears to only have 1 significant digit in one case, and being able to get full precision is useful in general.

**Comment by knelli2** [comment]
@wthrowe Like this?

```
Time                   NumberOfPoints         Volume                 L2Norm(GaugeConstraint) L2Norm(ThreeIndexConstraint)
2.0000000000000000e-02 1.5435000000000000e+05 4.1887614885579874e+06 5.7621831920527475e-08  2.4807265343400288e-06      
```

Edit: I set the default to be 16.

**Review by wthrowe — APPROVED** [review]
Yes, that's fine.  Thanks!

**Review by nilsvu — APPROVED** [review]

**Pull request merged**